runtime/lua/vim/lsp/buf.lua

```diff
@@ -249,13 +249,34 @@ end
 ---@param new_name (string) If not provided, the user will be prompted for a new
 ---name using |input()|.
 function M.rename(new_name)
-  -- TODO(ashkan) use prepareRename
-  -- * result: [`Range`](#range) \| `{ range: Range, placeholder: string }` \| `null` describing the range of the string to rename and optionally a placeholder text of the string content to be renamed. If `null` is returned then it is deemed that a 'textDocument/rename' request is not valid at the given position.
   local params = util.make_position_params()
-  new_name = new_name or npcall(vfn.input, "New Name: ", vfn.expand('<cword>'))
-  if not (new_name and #new_name > 0) then return end
-  params.newName = new_name
-  request('textDocument/rename', params)
+  local function prepare_rename(err, result)
+    if err == nil and result == nil then
+      print('nothing to rename')
+      return
+    end
+    if result ~= nil and result.placeholder then
+      new_name = new_name or npcall(vfn.input, "New Name: ", result.placeholder)
+    elseif result ~= nil and result.start and result['end'] and
+      result.start.line == result['end'].line then
+      local line = vfn.getline(result.start.line+1)
```

Comment:
Might be better to use nvim_buf_get_lines instead of calling via vimscript.

Reply:
i'll leave it as is :)